src/Composer/Util/Platform.php

```diff
@@ -56,13 +56,19 @@ public static function getCwd(bool $allowEmpty = false): string
 
     /**
      * Infallible realpath version that falls back on the given $path if realpath is not working
+     *
+     * NB: Do not remove recursive `realpath()` calls. Fixes failing Windows `realpath()` implementation.
+     * @see https://bugs.php.net/bug.php?id=72738
      */
     public static function realpath(string $path): string
     {
         $realPath = realpath($path);
         if ($realPath === false) {
             return $path;
```

Comment:
```suggestion
            return file_exists($realPath) || is_dir($realpath);
```

I was applying the `realpath()` to `Platform::realpath()` change across the other cases, and they are sometimes followed by `if($realpathResult === false){ }` blocks.  E.g.

https://github.com/composer/composer/blob/85ff84d6c5260ba21740a7c5c9a111890805d6e7/src/Composer/Package/Archiver/ArchivableFilesFinder.php#L50-L53

I'm thinking `Platform::realpath()` should return `false` if the path is incorrect, which is inherent in a `realpath('/does/not/exist')` call, and which the stream wrapper should handle correctly.

Reply:
> I'm thinking `Platform::realpath()` should return `false` if the path is incorrect, which is inherent in a `realpath('/does/not/exist')` call, and which the stream wrapper should handle correctly.

Yes, this complicates Platform::realpath() usage.. Ideally we avoid returning false, because checking for it everywhere is a pain. I can see with stream wrappers it's probably more likely to return false, but what are conditions where they would do that? Can we recover from this or should we just throw?

The fallback to `return $path;` that we currently have in Platform::realpath is because in theory realpath shouldn't fail when the path exists, and we don't call it if the path doesn't exist AFAIK. I have never been able to conclusively say what might trigger a false otherwise, so I put this fallback in place and nobody ever complained so either it isn't used or it works 🤷🏻‍♂️ . 

I'd probably be fine throwing in case false is returned, and then we see what happens in the real world, if it causes issues we can still revert to falling back for non-stream-wrapper paths I guess to restore the previous behavior.

Stream wrapper usage is definitely very fringe, so I am a bit worried of degrading the overall experience just to fix this.

> Any suggestion on writing tests on this? I often use [antecedent/patchwork](https://github.com/antecedent/patchwork) which allows redefining built in PHP functions. Although I see [composer/composer is using PHPUnit](https://github.com/composer/composer/blob/002114b7407d3b6fabcb041b28379ce416405cdc/composer.json#L48) in an unconventional way.

You should be able to use patchwork as it requires php 7.1+, so just give it a shot I guess..